src/passes/TranslateEH.cpp

```diff
@@ -0,0 +1,823 @@
+  // reused in both trys.
+  //
+  // Also even if there is a try-catch nest with depth N, if only some of them
+  // is targetted by rethrows, we only need that many extra local for the whole
+  // nest:
+  // (try $l0
+  //   (catch
+  //     (try $l1
+  //       (catch
+  //         ...
+  //         ...
+  //         (try $lN
+  //           (catch
+  //             ;; If this is the only rethrow in this nest, we only need one
+  //             ;; extra exnref local for all N try-catches
+  //             (rethrow $l1)
+  //           )
+  //         )
+  //         ...
+  //         ...
+  //       )
+  //     )
+  //   )
+  // )
+  struct ExnrefLocalAssigner : public PostWalker<ExnrefLocalAssigner> {
+    TargetTryLabelScanner* labelScanner = nullptr;
+
+    std::vector<Index> exnrefLocals;
+    std::unordered_map<Name, Index> rethrowTargetToExnrefLocal;
+
+    // Depth of the current try nest, when only counting trys targetted with
+    // rethrows.
+    size_t rethrowTryDepth = 0;
+
+    ExnrefLocalAssigner(Function* func, TargetTryLabelScanner* labelScanner)
+      : labelScanner(labelScanner) {
+      walkFunction(func);
+    }
+
+    bool hasExnrefLocal(Name rethrowTarget) const {
+      return rethrowTargetToExnrefLocal.find(rethrowTarget) !=
+             rethrowTargetToExnrefLocal.end();
+    }
+    Index getExnrefLocal(Name rethrowTarget) const {
+      auto it = rethrowTargetToExnrefLocal.find(rethrowTarget);
+      assert(it != rethrowTargetToExnrefLocal.end());
+      return it->second;
+    }
+
+    static void incrementRethrowTryDepth(ExnrefLocalAssigner* self,
+                                         Expression** currp) {
+      self->rethrowTryDepth++;
+    }
+    static void decrementRethrowTryDepth(ExnrefLocalAssigner* self,
+                                         Expression** currp) {
+      self->rethrowTryDepth--;
+    }
+
+    static void scan(ExnrefLocalAssigner* self, Expression** currp) {
+      auto* curr = *currp;
+      if (auto* tryy = curr->dynCast<Try>()) {
+        if (self->labelScanner->isTargettedByRethrows(tryy)) {
+          self->pushTask(decrementRethrowTryDepth, currp);
+        }
+      }
+      PostWalker<ExnrefLocalAssigner>::scan(self, currp);
+      if (auto* tryy = curr->dynCast<Try>()) {
+        if (self->labelScanner->isTargettedByRethrows(tryy)) {
+          self->pushTask(incrementRethrowTryDepth, currp);
+        }
+      }
+    }
+
+    void visitTry(Try* curr) {
+      if (labelScanner->isTargettedByRethrows(curr)) {
+        while (exnrefLocals.size() < rethrowTryDepth) {
+          exnrefLocals.push_back(
+            Builder::addVar(getFunction(), Type(HeapType::exn, Nullable)));
+        }
+        rethrowTargetToExnrefLocal[curr->name] =
+          exnrefLocals[rethrowTryDepth - 1];
+      }
+    }
+  };
+
+  std::unique_ptr<LabelUtils::LabelManager> labels;
+  std::unique_ptr<TargetTryLabelScanner> labelScanner;
+  std::unique_ptr<ExnrefLocalAssigner> localAssigner;
```

Comment:
I used `unique_ptr` mainly to avoid manually deleting those instances. Does `std::optional` delete the pointers for me? I don't think so... Would there be a better way to do this?

Reply:
I mean you could use e.g. `std::optional<LabelUtils::LabelManager> labels;`. It would still manage the lifetime of the object for you automatically and let you initialize it whenever you want, it just wouldn't require a heap allocation.

Reply:
Ah, `std::optional` also deletes the pointer when it is deleted. But in this case they are not strictly optional; they are just initialized later. Given that `std::optional` also deletes the pointer, I will change the code to use it, but what's a good practice, or criteria, on how to use `std::unique_ptr` vs. `std::optional`?

Reply:
https://github.com/WebAssembly/binaryen/pull/6210/commits/87f65d34d07282351868d39c7fff9390b7beac00

Reply:
I would say that std::optional is always better unless there’s a specific reason you need to do a heap allocation, e.g. very large data that would overflow the stack if stored in a std::optional or recursive data that requires an indirection to avoid having infinite size.